File: LeetCode/python3/647.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        total = len(s)
        dp = [[0 for i in range(len(s))] for j in range(len(s))]
        for i in range(len(s)):
            dp[i][i] = 1
        for i in range(len(s)):
            for j in range(i):
                if s[i] == s[j] and (i == j + 1 or dp[j + 1][i - 1] == 1):
                    total += 1
                    dp[j][i] = 1
        return total

File: LeetCode/cpp/647.cpp
class Solution {
public:
    int countSubstrings(string s) {
        int size = s.size();
        int count = 0;
        vector<vector<bool>> dp(size, vector<bool>(size, false));
        for (int i = 0; i < size; ++i) {
            for (int j = 0; j <= i; ++j) {
                dp[j][i] = ((i == j) || (s[i] == s[j] && ((i <= j + 2) || (dp[j + 1][i - 1] == 1))));
                if (dp[j][i])
                    ++count;
            }
        }
        return count;
    }
};